nameserver/server.go

```diff
@@ -21,59 +23,95 @@ func makeDNSFailResponse(r *dns.Msg) *dns.Msg {
 	return m
 }
 
-type DNSServer struct {
-	config *dns.ClientConfig
-	zone   Zone
-	iface  *net.Interface
-	port   int
-	udpSrv *dns.Server
-	tcpSrv *dns.Server
+type DNSServerConfig struct {
+	// (Optional) client config file for resolving upstream servers
+	UpstreamCfgFile string
+	// (Optional) DNS client config for the fallback server(s)
+	UpstreamCfg *dns.ClientConfig
+	// DNS zone served
+	Zone Zone
+	// Interface where serving
+	Iface *net.Interface
+	// (Optional) port number (for TCP and UDP)
+	Port int
+	// (Optional) local domain (ie, "weave.local.")
+	LocalDomain string
 }
 
-// Creates a new DNS server with a given config
-func NewDNSServer(config *dns.ClientConfig, zone Zone, iface *net.Interface, port int) *DNSServer {
-	return &DNSServer{
-		config: config,
-		zone:   zone,
-		iface:  iface,
-		port:   port,
-	}
+type DNSServer struct {
+	config  *DNSServerConfig
+	udpSrv  *dns.Server
+	tcpSrv  *dns.Server
+	mdnsCli *MDNSClient
+	mdnsSrv *MDNSServer
+
+	IfaceName  string // the interface where mDNS is working on
+	ListenAddr string // the address the server is listening at
 }
 
-// Start the DNS server
-func (s *DNSServer) Start() error {
-	Info.Printf("WeaveDNS server starting...")
+// Creates a new DNS server
+func NewDNSServer(config DNSServerConfig) (s *DNSServer, err error) {
+	// fill empty parameters with defaults...
+	if config.Port == 0 {
+		config.Port = DEFAULT_SERVER_PORT
+	}
+	if len(config.LocalDomain) == 0 {
+		config.LocalDomain = DEFAULT_LOCAL_DOMAIN
+	}
```

Comment:
I assume there are no restrictions in the local domain name. Should we check that the `LocalDomain` follows some rules (ie, it has `"local."` at the end)?